### ISSUE TITLE
Misleading warning about argument not being used

### ISSUE THREAD

**Issue by IamGianluca** [rust-lang/rust#116747]
Hi,

I've created a repository class that can retrieve the password for a user by passing the `user_id`. 

```rust
    fn get_password_by_user_id(&mut self, user_id: i32) -> Option<Password> {
        use crate::adapters::schema::credentials::dsl::*;

        credentials
            .filter(user_id.eq(user_id))
            .first::<Password>(&mut *self.conn)
            .ok()
    }
```

The corresponding schema for the `credentials` table also contains a field called `user_id`. 

```rust
diesel::table! {
    credentials (id) {
        id -> Int4,
        user_id -> Int4,
        password -> Text,
        created_at -> Timestamptz,
    }
}
```

The code executes correctly, but I get a warning which is misleading, as the field is actually being used. I have unit tests validating it. 

```
cargo test -- --test-threads=1
   Compiling rusty v0.1.0 (/home/gianluca/git/rusty)
warning: unused variable: `user_id`
  --> src/adapters/user_repository.rs:53:43
   |
53 |     fn get_password_by_user_id(&mut self, user_id: i32) -> Option<Password> {
   |                                           ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_user_id`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: `rusty` (bin "rusty" test) generated 1 warning (run `cargo fix --bin "rusty" --tests` to apply 1 suggestion)
    Finished test [unoptimized + debuginfo] target(s) in 2.08s
     Running unittests src/main.rs (target/debug/deps/rusty-22e64a5159198c30)
```

The language server also gets this wrong.

<img width="1616" alt="image" src="https://github.com/rust-lang/rust/assets/4025968/7d597584-313f-4e93-9718-da7f74e40b51">

**Comment by fmease** [comment]
Are you positive that the `user_id` inside the function body refers to the function parameter?
It seems like the module `dsl` exports an item `user_id` which gets added to the scope through the glob import. I don't know what `diesel::table` expands to, it might a top-level item named `user_id`. Consider this scenario:

```rs
mod dsl {
    #[allow(non_upper_case_globals)]
    pub const user_id: &i32 = &0;
}

pub fn get(user_id: i32) { //~ WARN unused variable: `user_id`
    use dsl::*;

    let _ = user_id.eq(user_id);
}

```

**Comment by IamGianluca** [comment]
You're right! I should have been more thoughtful when creating the test. As you suggested, it seems that the `dsl` exports a `user_id` variable with a default value of `1`. I had to test the function with a value different from `1` and validated my understanding was wrong. Apologies for the false alarm! 😓

**Comment by fmease** [comment]
No worries, glad I could help :) At some point in the future, rustc could maybe emit a warning in this case, although I'm not sure how easy it'd be to detect.
